apps/labs/posts/integrating-hypothesis-into-sympy.md

```diff
@@ -0,0 +1,103 @@
+```
+
+# What is Property Based Testing?
+
+Property based testing (PBT) is a technique where instead of testing individual test cases, you specify properties you wish to hold true for a range of inputs. These properties are then tested against automatically generated test data. PBT uses logical properties over generated test data to facilitate broad, robust testing that can expose edge cases not easily found via traditional test cases.
+
+PBT _shines_ when testing generic functions and libraries working across a wide range of possible inputs, where manually enumerating test cases is difficult but the behavior in question is easily testable. Examples include:
+
+- Mathematical Operations
+- String formatting
+- Database migration
+- Compression algorithms
+
+Let's take a simple, concrete example. Say we want to ensure that after multiplying two polynomials together, the degree of the resulting polynomial is the sum of the degrees of the two polynomials. We could write a test case for this:
+
+```python
+from sympy.abc import x
+def test_degree():
+    f = Poly(x**2 + 1, x)
+    g = Poly(x**3 + 1, x)
+    h = f*g
+    assert h.degree() == f.degree() + g.degree()
+```
+
+Notice we are limited in how many various combinations of _f_ and _g_ we can test. It would be better if we could fix a property and have a library automatically generate interesting test cases and run them for us. We would no longer need to worry about thinking up input/output pairs. This would increase our **_trust_** in the implementation.
+
+# What is Hypothesis?
+
+> [Hypothesis](https://hypothesis.readthedocs.io/en/latest/) is a Python library for creating unit tests which are simpler to write and more powerful when run, finding edge cases in your code you wouldn’t have thought to look for. It is stable, powerful and easy to add to any existing test suite.
+> Now, let's test out the property above using Hypothesis:
+
+```python
+from hypothesis import given, strategies as st
+
+@given(f = polys(), g = polys())
+def test_degree(f, g):
+    h = f * g
+    assert h.degree() == f.degree() + g.degree()
+```
+
+Note, that here `polys()` is custom-built for sympy and generates a random polynomial with integer coefficients. It is not a built-in Hypothesis strategy.
+
+### How Hypothesis Works
+
+Give Hypothesis the types of inputs you are expecting using the _@given_ decorator and it will automatically generate examples using the _strategies_ module. It will then use these examples to test and report the minimal failing inputs (if any).
+
+Hypothesis is able to come up with interesting inputs using a combination of smart generation, guiding metrics, and feedback loops. For example, the built-in stategies like `integers()` have default behaviors tuned for common useful values. That is, while some inputs are random, it also tries to choose cases that commonly cause errors (like 0 or NaN). Hypothesis is _adversarial_ in the way it chooses inputs to test against your function.
```

Comment:
```suggestion
Hypothesis is able to come up with interesting inputs using a combination of smart generation, guiding metrics, and feedback loops. For example, the built-in strategies like `floats()` have default behaviors tuned for common useful values. That is, while some inputs are random, it also tries to choose cases that commonly cause errors (like 0 or NaN). Hypothesis is _adversarial_ in the way it chooses inputs to test against your function.
```

1. Grammar fix
2. `floats()` is the better example, prob what Diane was going for.